NIFI-12857 Add test for and refactor implementation of QueuePrioritizer

EndzeitBegins opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

Adds test for and refactors implementation of `QueuePrioritizer`.

[NIFI-12857](https://issues.apache.org/jira/browse/NIFI-12857)

Additionally, removed some dead code from `TestStandardFlowFileQueue`.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

exceptionfactory reviewed (commented):
@EndzeitBegins can you rebase this PR from the current main branch? It includes some improvements for the build process that should address the last failure in the Static Analysis workflow step.

EndzeitBegins commented:
Of course, just rebased on the last commit of `main`.

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/QueuePrioritizer.java:
```diff
@@ -18,74 +18,41 @@
-            }
-        }
+        final int claimComparisionResult = claimComparator.compare(f1, f2);
+        if (claimComparisionResult != 0) return claimComparisionResult;
```
As a general rule, braces should be used for conditional blocks:
```suggestion
        if (claimComparisionResult != 0) {
            return claimComparisionResult;
        }
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/QueuePrioritizer.java:
```diff
@@ -18,74 +18,41 @@
-            }
-        }
+        final int penaltyComparisonResult = penaltyComparator.compare(f1, f2);
+        if (penaltyComparisonResult != 0) return penaltyComparisonResult;
```
```suggestion
        if (penaltyComparisonResult != 0) {
            return penaltyComparisonResult;
        }
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/QueuePrioritizer.java:
```diff
@@ -18,74 +18,41 @@
-            }
+        for (FlowFilePrioritizer comparator : prioritizers) {
+            final int prioritizerComparisonResult = comparator.compare(f1, f2);
+            if (prioritizerComparisonResult != 0) return prioritizerComparisonResult;
```
```suggestion
            if (prioritizerComparisonResult != 0) {
                return prioritizerComparisonResult;
            }
```

exceptionfactory commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/test/java/org/apache/nifi/controller/queue/QueuePrioritizerTest.java:
```diff
@@ -0,0 +1,135 @@
+        assertEquals(0, prioritizer.compare(nonPenalizedFlowFile, nonPenalizedFlowFile));
+        assertEquals(1, prioritizer.compare(nonPenalizedFlowFile, expiredPenaltyFlowFile));
+        assertEquals(-1, prioritizer.compare(nonPenalizedFlowFile, penalizedFlowFile));
+        assertEquals(-1, prioritizer.compare(nonPenalizedFlowFile, longerPenalizedFlowFile));        assertEquals(1, prioritizer.compare(longerPenalizedFlowFile, nonPenalizedFlowFile));
```
```suggestion
        assertEquals(-1, prioritizer.compare(nonPenalizedFlowFile, longerPenalizedFlowFile));
        assertEquals(1, prioritizer.compare(longerPenalizedFlowFile, nonPenalizedFlowFile));
```

exceptionfactory reviewed (changes requested):
@EndzeitBegins Returning to this PR, I noticed a couple stylistic adjustments needed, but with those and one more rebase, this looks close to completion.

EndzeitBegins commented:
Thank you for the review @exceptionfactory. 

I've just rebased the PR on the last commit on `main`, added the missing line break to the test, and added braces to conditional blocks where they were missing. 

I would appreciate another look.

exceptionfactory reviewed (approved):
Thanks @EndzeitBegins! +1 merging
